fax/implicit/twophase.py

```diff
@@ -32,98 +29,44 @@ def convergence_test(x_new, x_old):
         )
 
         return sol
-    return _default_solver
+    return _default_solve
 
 
-def make_adjoint_fixed_point_iteration(
-        param_func, default_rtol=1e-4, default_atol=1e-4, default_max_iter=5000,
-        default_batched_iter_size=1):
+@functools.partial(jax.custom_vjp, nondiff_argnums=(0, 1, 3))
+def two_phase_solve(param_func, init_xs, params, solvers=()):
+    if solvers:
+        fwd_solver = solvers[0]
+    else:
+        fwd_solver = default_solver(param_func)
 
-    # define a flat function to fit the vjp API
-    def flat_func(i, x, params):
-        return param_func(params)(i, x)
+    return fwd_solver(init_xs, params)
 
-    def _adjoint_iteration_vjp(g, ans, init_xs, params):
-        dvalue = g
-        del init_xs
-        init_dxs = dvalue
 
-        fp_vjp_fn = jax.vjp(partial(flat_func, ans.iterations),
-                            ans.value, params)[1]
+def two_phase_fwd(param_func, init_xs, params, solvers):
+    sol = two_phase_solve(param_func, init_xs, params, solvers)
+    return sol, (sol.value, params)
```

Comment:
@pierrelux brought up offline a very good point. Assumptions about what is returned by the forward solver should be kept to an absolute minimum. Currently, it is assumed that a `FixedPointSolution` is returned. With the addition of `jax.experimental.host_callback`, this seems unnecessary and should be removed.

Reply:
@gehring I agree that your `FixedPointSolution` is an assumption.   I did the same thing with my implementation's [return type, `AugmentedState`](https://github.com/NeilGirdhar/fax_suggestions/blob/master/cmm/fixed_point/augmented.py).   Incidenetally, I'm just about to change that since with stochastic iterated functions, convergence isn't as simple as comparing the last state with the current state.  You need to roughly check that the variance of a set of recent states is small, which means keeping a different variety of statistics.

What does `host_callback` have to do with the return value of the fixed point solver?  You shouldn't be using it to return anything if you can avoid it.

Reply:
The original reason for having things set up this way was to be able to report/monitor the number of iterations used and evaluate the convergence. My understanding is that `host_callback` lets us extract that kind of info.

It would allow us to focus on handling the differentiation and make it easier for users to write their own functions/solvers. If someone wants to extract some internal quantities related to the solvers, they can make use of `host_callback` in a custom solver and use that instead of the default solver.

Reply:
Yes, one use is to report the number of iterations, but don't you also want to keep track of the number of iterations so that you can stop iterating when a maximum is reached?

I also used the host callback mechanism, but mainly to show a progress bar for long iterations.

Reply:
Also, using `host_callback` is probably not as fast as reporting various statistics using the regular XLA machinery.

Reply:
That can still be used internally. What I want to get rid of is the assumption that the solver returns the number of iterations used (as well as some other stuff). I want to assume that the solver only returns the solution, nothing else, similarly to what `jax.lax.custom_root` expects as its `solve` argument.

Reply:
> Also, using host_callback is probably not as fast as reporting various statistics using the regular XLA machinery.

It shouldn't be used extensively. If the goal is to log the number of iterations used, you would only require one call for every time you solve a fixedpoint. Unless `host_callback` is extremely slow, I would expect that solving the fixed point will dominate your runtime. That being said, it is probably safe to measure the overhead before committing.

Reply:
Okay, I'll be curious to see your final design.  I'll have my redesign done this week if you're curious.

Reply:
I'll look into this a bit later (#28)